libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1890,52 +1923,86 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+			/// For error message: Struct members that were removed during conversion to memory.
+			TypePointer const expressionType = type(_functionCall.expression());
+			TypeType const& t = dynamic_cast<TypeType const&>(*expressionType);
+			auto const& structType = dynamic_cast<StructType const&>(*t.actualType());
+			set<string> membersRemovedForStructConstructor = structType.membersMissingInMemory();
+
+			if (!membersRemovedForStructConstructor.empty())
+			{
+				msg += " Members that have to be skipped in memory:";
+				for (auto const& member: membersRemovedForStructConstructor)
+					msg += " " + member;
+			}
 		}
 		else if (
-			functionType->kind() == FunctionType::Kind::BareCall ||
-			functionType->kind() == FunctionType::Kind::BareCallCode ||
-			functionType->kind() == FunctionType::Kind::BareDelegateCall ||
-			functionType->kind() == FunctionType::Kind::BareStaticCall
+			_functionType->kind() == FunctionType::Kind::BareCall ||
+			_functionType->kind() == FunctionType::Kind::BareCallCode ||
+			_functionType->kind() == FunctionType::Kind::BareDelegateCall ||
+			_functionType->kind() == FunctionType::Kind::BareStaticCall
 		)
 		{
 			if (arguments.empty())
-				msg += " This function requires a single bytes argument. Use \"\" as argument to provide empty calldata.";
+				msg +=
+					" This function requires a single bytes argument."
+					" Use \"\" as argument to provide empty calldata.";
 			else
-				msg += " This function requires a single bytes argument. If all your arguments are value types, you can use abi.encode(...) to properly generate it.";
+				msg +=
+					" This function requires a single bytes argument."
+					" If all your arguments are value types, you can use"
+					" abi.encode(...) to properly generate it.";
 		}
 		else if (
-			functionType->kind() == FunctionType::Kind::KECCAK256 ||
-			functionType->kind() == FunctionType::Kind::SHA256 ||
-			functionType->kind() == FunctionType::Kind::RIPEMD160
+			_functionType->kind() == FunctionType::Kind::KECCAK256 ||
+			_functionType->kind() == FunctionType::Kind::SHA256 ||
+			_functionType->kind() == FunctionType::Kind::RIPEMD160
 		)
 			msg +=
 				" This function requires a single bytes argument."
-				" Use abi.encodePacked(...) to obtain the pre-0.5.0 behaviour"
-				" or abi.encode(...) to use ABI encoding.";
+				" Use abi.encodePacked(...) to obtain the pre-0.5.0"
+				" behaviour or abi.encode(...) to use ABI encoding.";
 		m_errorReporter.typeError(_functionCall.location(), msg);
+		return;
 	}
-	else if (isPositionalCall)
+
+	if (isPositionalCall)
 	{
+		bool const abiEncoderV2 = m_scope->sourceUnit().annotation().experimentalFeatures.count(
+			ExperimentalFeature::ABIEncoderV2
+		);
+
 		for (size_t i = 0; i < arguments.size(); ++i)
 		{
 			auto const& argType = type(*arguments[i]);
-			if (functionType->takesArbitraryParameters() && i >= parameterTypes.size())
+			if (_functionType->takesArbitraryParameters() && i >= parameterTypes.size())
```

Comment:
I just re-checked: The only functions where `takesArbitraryParameters()` returns true are the encoding and decoding functions. Perhaps it would make sense to factor them out into a single checking function?

Reply:
Argument type/count checking is now in a standalone method which handles additional arguments (i.e. arbitrary arguments beyond the function specified parameters). As the arbitrary parameters are not used in any other checking, I don't think it makes sense to separate out encoding functions for that reason alone.